core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java

```diff
@@ -0,0 +1,159 @@
+ *    }
+ * </pre>
+ * <p>
+ * Files stored in `sfs/corrupt` contain already corrupted .dat files that were originally created in 11.x and were
+ * migrated to 12.1.x with the broken SFS migration code. Hence, when these .dat files are loaded the SFS attempts to
+ * recover the corrupt data.
+ *
+ * @author Ryan Emerson
+ * @since 13.0
+ */
+@Test(groups = "unit", testName = "persistence.file.SingleFileGracefulShutdownMigrationTest")
+public class SingleFileGracefulShutdownMigrationTest extends AbstractInfinispanTest {
+
+   private static final String CACHE_NAME = "update-cache";
+
+   private String tmpDirectory;
+
+   @BeforeClass
+   protected void setUpTempDir() throws IOException {
+      tmpDirectory = CommonsTestingUtil.tmpDirectory(this.getClass());
+      new File(tmpDirectory).mkdirs();
+   }
+
+   @AfterClass
+   protected void clearTempDir() {
+      Util.recursiveFileRemove(tmpDirectory);
+   }
+
+   enum Marshaller {
+      PROTOSTREAM,
+      JAVA_SERIALIZATION
+   }
+
+   @DataProvider(name = "testFiles")
+   Object[][] singleTypes() {
+      return new Object[][]{
+            // Not possible to test corrupt Marshaller.JAVA_SERIALIZATION bytes as previously 11.x -> 12. migration always fails with non-protostream marshallers
+            // MAGIC_11_0
+            {"sfs/11_0/all-entries-java-serialization.dat", Marshaller.JAVA_SERIALIZATION, 1003},
+            // MAGIC_12_0 created on 12.1.x
+            {"sfs/12_0/all-entries-java-serialization.dat", Marshaller.JAVA_SERIALIZATION, 1003},
+            // 11.x -> 12.1.x migration resulting in corrupt data
+            {"sfs/corrupt/all-entries.dat", Marshaller.PROTOSTREAM, 2003},
+            // MAGIC_12_0 created on 12.1.x
+            {"sfs/12_0/all-entries.dat", Marshaller.PROTOSTREAM, 1003},
+      };
+   }
+
+   @Test(dataProvider = "testFiles")
+   public void testAllEntriesRecovered(String fileName, Marshaller marshallerType, int numEntries) throws Exception {
+      InputStream is = FileLookupFactory.newInstance().lookupFile(fileName, Thread.currentThread().getContextClassLoader());
+      Files.copy(is, Paths.get(tmpDirectory).resolve(CACHE_NAME + ".dat"), StandardCopyOption.REPLACE_EXISTING);
+
+      ConfigurationBuilderHolder cbh = new ConfigurationBuilderHolder();
+      if (marshallerType == Marshaller.JAVA_SERIALIZATION)
+         cbh.getGlobalConfigurationBuilder().serialization().marshaller(new JavaSerializationMarshaller());
+
+      cbh.newConfigurationBuilder(CACHE_NAME)
+            .persistence()
+            .addSingleFileStore()
+            .segmented(false)
+            .location(tmpDirectory);
+
+      try (EmbeddedCacheManager cacheManager = new DefaultCacheManager(cbh, true)) {
+         Cache<Object, Object> cache = cacheManager.getCache(CACHE_NAME);
+         ByRef.Integer i = new ByRef.Integer(0);
+         // Iterate all entries to ensure values can be read
+         cache.forEach((k, v) -> i.inc());
+         assertEquals(numEntries, i.get());
```

Comment:
I would add 2 writes here. 
* write to a migrated key
* write to a new key

below, check if both are still there too :)